backend/app.py

```diff
@@ -1,49 +1,146 @@
-CORS(app, resources={r"/*":{"origins":"*"}})
+app.config['SECRET'] = environ.get('API_TOKEN')
+app.config['SQLALCHEMY_DATABASE_URI'] = environ.get('CONN_STR')
+db = SQLAlchemy(app)
+CORS(app, resources={r"/*": {"origins": "*"}})
 socketio = SocketIO(app, cors_allowed_origins="*")
 
+
+def authenticate(func):
+    """Wrapper function to authenticate routes"""
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        token = None
+        if 'x-access-token' in request.headers:
+            token = request.headers['x-access-token']
+        if not token:
+            return jsonify({'message': 'You must be logged in to access this page'}), 401
+        try:
+            data = jwt.decode(
+                token, app.config['SECRET'], algorithms=["HS256"])
+            current_user = db.session.execute(
+                db.select(User).filter_by(id=data['id'])).one()
+        except (jwt.InvalidIssuedAtError, jwt.ExpiredSignatureError, exc.SQLAlchemyError):
+            return jsonify({'message': 'Invalid Login'}), 401
+        return func(current_user, *args, **kwargs)
+    return wrapper
+
+
 @socketio.on("connect")
 def on_connect():
     """listener for connection events"""
     print("client has connected " + request.sid)
-    emit("connect",{ "data":f"id: {request.sid} is connected"})
+    emit("connect", {"data": f"id: {request.sid} is connected"})
+
 
 @socketio.on("disconnect")
 def on_disconnect():
     """listener for disconnect events"""
     print("client has disconnected " + request.sid)
-    emit("disconnect", {"data":f"id: {request.sid} has disconnected"}, broadcast=True)
+    emit("disconnect", {
+         "data": f"id: {request.sid} has disconnected"}, broadcast=True)
+
 
 @socketio.on('data')
 def handle_message(data):
     """listener for data events"""
-    print("data from the front end: ",str(data))
-    emit("data", {'data':data,'id':request.sid}, broadcast=True)
+    print("data from the front end: ", str(data))
+    emit("data", {'data': data, 'id': request.sid}, broadcast=True)
+
 
 @app.route("/")
 def root():
     """returns object with filler data"""
-    data = {'data':'Root accessed. Secret token is ' + environ.get('api-token')}
+    ##################################################
+    # TEMP TO CLEAR USERS AS MODEL IS BEING UPDATED
+    ##################################################
+    # This deletes all the users from the Users table
+    # try:
+    #     db.session.query(User).delete()
+    #     db.session.commit()
+    #     print("users deleted")
+    # except:
+    #     db.session.rollback()
+    #     print("users not deleted")
+    ##################################################
+    # END TEMP
+    ##################################################
+
+    data = {'data': 'Root accessed. Secret token is ' + app.config['SECRET']}
     return jsonify(data)
 
-@app.route('/route_name', methods=['GET', 'POST'])
-def method_name():
+
+@app.route('/user')
+@authenticate
```

Comment:
What does the `@authenticate` decorator do? Can't seem to find any documentation for it.

Reply:
It's actually defined a little further up on line 25 and uses the `functools.wraps` function